### ISSUE TITLE
ROCm: DeepCopy Bug (device to device)

### ISSUE THREAD

**Issue by crtrott** [kokkos/kokkos#964]
The following code fails:

```C++
#include<Kokkos_Core.hpp>

int main(int argc, char* argv[]) {
  Kokkos::initialize(argc,argv);

  Kokkos::View<int> a("A");
  Kokkos::View<int> b("B");

  Kokkos::View<int>::HostMirror h_a = Kokkos::create_mirror_view(a);
  Kokkos::View<int>::HostMirror h_b = Kokkos::create_mirror_view(b);

  h_a() = 1;
  Kokkos::deep_copy(a,h_a);
  Kokkos::deep_copy(b,a);

  Kokkos::deep_copy(h_b,b);
  printf("%i %i\n",h_a(),h_b()); // Wrong answer here ("1 0")

  Kokkos::deep_copy(h_b,a);
  printf("%i %i\n",h_a(),h_b()); // Right answer here ("1 1")

  Kokkos::finalize();
}
```

**Comment by crtrott** [comment]
Upon further study it looks like the last 32 elements are not copied, consider this code with 1D-Views:
```C++
int main(int argc, char* argv[]) {
  Kokkos::initialize(argc,argv);

  int N = 35;
  Kokkos::View<int*> a("A",N);
  Kokkos::View<int*> b("B",N);

  Kokkos::View<int*>::HostMirror h_a = Kokkos::create_mirror_view(a);
  Kokkos::View<int*>::HostMirror h_b = Kokkos::create_mirror_view(b);

  for(int i=0; i<N; i++)
    h_a(i) = i+1;
  Kokkos::deep_copy(a,h_a);
  Kokkos::deep_copy(b,a);
  Kokkos::deep_copy(h_b,b);

  for(int i=0; i<N; i++)
    printf("%i %i\n",h_a(i),h_b(i));

  Kokkos::deep_copy(h_b,a);
  for(int i=0; i<N; i++)
    printf("%i %i\n",h_a(i),h_b(i));
  Kokkos::finalize();
}
```

This puts out:
```
1 1
2 2
3 3
4 0
5 0
6 0
7 0
8 0
9 0
10 0
11 0
12 0
13 0
14 0
15 0
16 0
17 0
18 0
19 0
20 0
21 0
22 0
23 0
24 0
25 0
26 0
27 0
28 0
29 0
30 0
31 0
32 0
33 0
34 0
35 0
1 1
2 2
3 3
4 4
5 5
6 6
7 7
8 8
9 9
10 10
11 11
12 12
13 13
14 14
15 15
16 16
17 17
18 18
19 19
20 20
21 21
22 22
23 23
24 24
25 25
26 26
27 27
28 28
29 29
30 30
31 31
32 32
33 33
34 34
35 35
```

**Comment by crtrott** [comment]
Using 92c8baba028e6b2e7af2b73d1b2b5ee0a7e8ab68

**Comment by chipfreitag** [comment]
I was able to build a standalone program(no kokkos involved), duplicating the issue.
The View SharedAlloc constructor puts data at an offset of +128 bytes from the start of allocated GPU memory.  That is just how much is missing.  The standalone example shows that the missing data tracks with the amount of the offset. 
Working this issue with the compiler team.

**Comment by chipfreitag** [comment]
Fixes made in the compiler/runtime.  Will advise when new version is available.

**Comment by chipfreitag** [comment]
Resolved ROCm 1.6, December release iirc.
